# Importing libraries

In [166]:
import pandas as pd
import numpy as np

# Load the data

In [167]:
book_ratings_df = pd.read_csv("BX-Book-Ratings.csv", sep = ';', on_bad_lines="skip", encoding='latin-1')

In [168]:
books_df = pd.read_csv("BX-Books.csv", sep = ';', on_bad_lines='skip', encoding='latin-1',low_memory=False)

In [169]:
users_df = pd.read_csv("BX-Users.csv", sep = ';', on_bad_lines='skip', encoding='latin-1')

In [170]:
book_ratings_df.head()

,User-ID,ISBN,Book-Rating
0,276725,034545104X,0
1,276726,0155061224,5
2,276727,0446520802,0
3,276729,052165615X,3
4,276729,0521795028,6


In [171]:
book_ratings_df.rename(columns={'User-ID':'user_id',
                             'ISBN' : 'isbn',
                             'Book-Rating' : 'rating'}, inplace=True)

In [172]:
book_ratings_df.head()

,user_id,isbn,rating
0,276725,034545104X,0
1,276726,0155061224,5
2,276727,0446520802,0
3,276729,052165615X,3
4,276729,0521795028,6


In [173]:
books_df.head()

,ISBN,Book-Title,Book-Author,Year-Of-Publication,Publisher,Image-URL-S,Image-URL-M,Image-URL-L
0,0195153448,Classical Mythology,Mark P. O. Morford,2002,Oxford University Press,http://images.amazon.com/images/P/0195153448.0...,http://images.amazon.com/images/P/0195153448.0...,http://images.amazon.com/images/P/0195153448.0...
1,0002005018,Clara Callan,Richard Bruce Wright,2001,HarperFlamingo Canada,http://images.amazon.com/images/P/0002005018.0...,http://images.amazon.com/images/P/0002005018.0...,http://images.amazon.com/images/P/0002005018.0...
2,0060973129,Decision in Normandy,Carlo D'Este,1991,HarperPerennial,http://images.amazon.com/images/P/0060973129.0...,http://images.amazon.com/images/P/0060973129.0...,http://images.amazon.com/images/P/0060973129.0...
3,0374157065,Flu: The Story of the Great Influenza Pandemic...,Gina Bari Kolata,1999,Farrar Straus Giroux,http://images.amazon.com/images/P/0374157065.0...,http://images.amazon.com/images/P/0374157065.0...,http://images.amazon.com/images/P/0374157065.0...
4,0393045218,The Mummies of Urumchi,E. J. W. Barber,1999,W. W. Norton &amp; Company,http://images.amazon.com/images/P/0393045218.0...,http://images.amazon.com/images/P/0393045218.0...,http://images.amazon.com/images/P/0393045218.0...


In [174]:
books_df = books_df[['ISBN', 'Book-Title', 'Book-Author','Year-Of-Publication', 'Publisher', 'Image-URL-L']]

In [175]:
books_df = books_df.rename(columns = {'ISBN':'isbn', 'Book-Title':'title', 
'Book-Author': 'author', 'Year-Of-Publication': 'year_of_publication',
'Publisher':'publisher',
'Image-URL-L':'image_url_L'})

In [176]:
books_df.head()

,isbn,title,author,year_of_publication,publisher,image_url_L
0,0195153448,Classical Mythology,Mark P. O. Morford,2002,Oxford University Press,http://images.amazon.com/images/P/0195153448.0...
1,0002005018,Clara Callan,Richard Bruce Wright,2001,HarperFlamingo Canada,http://images.amazon.com/images/P/0002005018.0...
2,0060973129,Decision in Normandy,Carlo D'Este,1991,HarperPerennial,http://images.amazon.com/images/P/0060973129.0...
3,0374157065,Flu: The Story of the Great Influenza Pandemic...,Gina Bari Kolata,1999,Farrar Straus Giroux,http://images.amazon.com/images/P/0374157065.0...
4,0393045218,The Mummies of Urumchi,E. J. W. Barber,1999,W. W. Norton &amp; Company,http://images.amazon.com/images/P/0393045218.0...


In [177]:
users_df.head()

,User-ID,Location,Age
0,1,"nyc, new york, usa",NaN
1,2,"stockton, california, usa",18.0
2,3,"moscow, yukon territory, russia",NaN
3,4,"porto, v.n.gaia, portugal",17.0
4,5,"farnborough, hants, united kingdom",NaN


In [178]:
users_df.rename(columns=
                {'User-ID' : 'user_id',
                 'Location': 'location',
                 'Age' : 'age'}, inplace=True)

In [179]:
users_df.head()

,user_id,location,age
0,1,"nyc, new york, usa",NaN
1,2,"stockton, california, usa",18.0
2,3,"moscow, yukon territory, russia",NaN
3,4,"porto, v.n.gaia, portugal",17.0
4,5,"farnborough, hants, united kingdom",NaN


In [180]:
book_ratings_df.shape

(1149780, 3)

In [181]:
books_df.shape

(271360, 6)

In [182]:
users_df.shape

(278858, 3)

In [183]:
book_ratings_df.isnull().sum()

user_id    0
isbn       0
rating     0
dtype: int64

In [184]:
books_df.isnull().sum()

isbn                   0
title                  0
author                 2
year_of_publication    0
publisher              2
image_url_L            3
dtype: int64

In [185]:
books_df[books_df['author'].isna() & books_df['image_url_L'].isna()]

,isbn,title,author,year_of_publication,publisher,image_url_L


In [186]:
books_df[books_df['author'].isna() & books_df['publisher'].isna()]

,isbn,title,author,year_of_publication,publisher,image_url_L


In [187]:
books_df.dropna(inplace=True)

In [188]:
books_df.isnull().sum()

isbn                   0
title                  0
author                 0
year_of_publication    0
publisher              0
image_url_L            0
dtype: int64

In [189]:
users_df.isnull().sum()

user_id          0
location         0
age         110762
dtype: int64

# Duplicate values

In [190]:
book_ratings_df.duplicated().sum()

np.int64(0)

In [191]:
books_df.duplicated().sum()

np.int64(0)

In [192]:
users_df.duplicated().sum()

np.int64(0)

# Exploratory Dta Analysis

In [193]:
user_rating_counts = book_ratings_df['user_id'].value_counts()
user_rating_counts

user_id
11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: count, Length: 105283, dtype: int64

In [194]:
active_users = user_rating_counts[user_rating_counts > 200]

In [195]:
active_users

user_id
11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
274808      201
28634       201
59727       201
268622      201
188951      201
Name: count, Length: 899, dtype: int64

In [196]:
active_users.shape

(899,)

In [197]:
filtered_ratings_df = book_ratings_df[
    book_ratings_df['user_id'].isin(active_users.index)]

In [198]:
filtered_ratings_df.head()

,user_id,isbn,rating
1456,277427,002542730X,10
1457,277427,0026217457,0
1458,277427,003008685X,8
1459,277427,0030615321,0
1460,277427,0060002050,0


# Merge ratings and books datasets

In [199]:
ratings_with_books = filtered_ratings_df.merge(books_df, on = 'isbn')

In [200]:
ratings_with_books.head()

,user_id,isbn,rating,title,author,year_of_publication,publisher,image_url_L
0,277427,002542730X,10,Politically Correct Bedtime Stories: Modern Ta...,James Finn Garner,1994,John Wiley &amp; Sons Inc,http://images.amazon.com/images/P/002542730X.0...
1,277427,0026217457,0,Vegetarian Times Complete Cookbook,Lucy Moll,1995,John Wiley &amp; Sons,http://images.amazon.com/images/P/0026217457.0...
2,277427,003008685X,8,Pioneers,James Fenimore Cooper,1974,Thomson Learning,http://images.amazon.com/images/P/003008685X.0...
3,277427,0030615321,0,"Ask for May, Settle for June (A Doonesbury book)",G. B. Trudeau,1982,Henry Holt &amp; Co,http://images.amazon.com/images/P/0030615321.0...
4,277427,0060002050,0,On a Wicked Dawn (Cynster Novels),Stephanie Laurens,2002,Avon Books,http://images.amazon.com/images/P/0060002050.0...


## Count ratings for each book

In [201]:
count_ratings = ratings_with_books.groupby('title')['rating'].count().reset_index()

In [202]:
count_ratings

,title,rating
0,A Light in the Storm: The Civil War Diary of ...,2
1,Always Have Popsicles,1
2,Apple Magic (The Collector's series),1
3,Beyond IBM: Leadership Marketing and Finance ...,1
4,Clifford Visita El Hospital (Clifford El Gran...,1
...,...,...
160259,Ã?Â?ber die Pflicht zum Ungehorsam gegen den S...,3
160260,Ã?Â?lpiraten.,1
160261,Ã?Â?rger mit Produkt X. Roman.,1
160262,Ã?Â?stlich der Berge.,1


In [203]:
count_ratings.rename(columns = {'rating' : 'no_of_ratings'}, inplace=True)

In [204]:
count_ratings

,title,no_of_ratings
0,A Light in the Storm: The Civil War Diary of ...,2
1,Always Have Popsicles,1
2,Apple Magic (The Collector's series),1
3,Beyond IBM: Leadership Marketing and Finance ...,1
4,Clifford Visita El Hospital (Clifford El Gran...,1
...,...,...
160259,Ã?Â?ber die Pflicht zum Ungehorsam gegen den S...,3
160260,Ã?Â?lpiraten.,1
160261,Ã?Â?rger mit Produkt X. Roman.,1
160262,Ã?Â?stlich der Berge.,1


In [205]:
final_ratings_df = ratings_with_books.merge(count_ratings, on='title')

In [206]:
final_ratings_df.head()

,user_id,isbn,rating,title,author,year_of_publication,publisher,image_url_L,no_of_ratings
0,277427,002542730X,10,Politically Correct Bedtime Stories: Modern Ta...,James Finn Garner,1994,John Wiley &amp; Sons Inc,http://images.amazon.com/images/P/002542730X.0...,82
1,277427,0026217457,0,Vegetarian Times Complete Cookbook,Lucy Moll,1995,John Wiley &amp; Sons,http://images.amazon.com/images/P/0026217457.0...,7
2,277427,003008685X,8,Pioneers,James Fenimore Cooper,1974,Thomson Learning,http://images.amazon.com/images/P/003008685X.0...,1
3,277427,0030615321,0,"Ask for May, Settle for June (A Doonesbury book)",G. B. Trudeau,1982,Henry Holt &amp; Co,http://images.amazon.com/images/P/0030615321.0...,1
4,277427,0060002050,0,On a Wicked Dawn (Cynster Novels),Stephanie Laurens,2002,Avon Books,http://images.amazon.com/images/P/0060002050.0...,13


## books with more than 50 number of ratings

In [207]:
final_ratings_df = final_ratings_df[final_ratings_df['no_of_ratings'] > 50]

In [208]:
final_ratings_df.head()

,user_id,isbn,rating,title,author,year_of_publication,publisher,image_url_L,no_of_ratings
0,277427,002542730X,10,Politically Correct Bedtime Stories: Modern Ta...,James Finn Garner,1994,John Wiley &amp; Sons Inc,http://images.amazon.com/images/P/002542730X.0...,82
13,277427,0060930535,0,The Poisonwood Bible: A Novel,Barbara Kingsolver,1999,Perennial,http://images.amazon.com/images/P/0060930535.0...,133
15,277427,0060934417,0,Bel Canto: A Novel,Ann Patchett,2002,Perennial,http://images.amazon.com/images/P/0060934417.0...,108
18,277427,0061009059,9,One for the Money (Stephanie Plum Novels (Pape...,Janet Evanovich,1995,HarperTorch,http://images.amazon.com/images/P/0061009059.0...,108
24,277427,006440188X,0,The Secret Garden,Frances Hodgson Burnett,1998,HarperTrophy,http://images.amazon.com/images/P/006440188X.0...,79


In [209]:
final_ratings_df.isnull().sum()

user_id                0
isbn                   0
rating                 0
title                  0
author                 0
year_of_publication    0
publisher              0
image_url_L            0
no_of_ratings          0
dtype: int64

In [210]:
final_ratings_df.shape

(59903, 9)

In [211]:
final_ratings_df.duplicated(subset=['user_id', 'title']).sum()

np.int64(1951)

In [212]:
dupes = final_ratings_df[final_ratings_df.duplicated(subset=['user_id','title'], keep=False)]
dupes.sort_values(['user_id','title']).head(10)

,user_id,isbn,rating,title,author,year_of_publication,publisher,image_url_L,no_of_ratings
5091,254,0439064864,9,Harry Potter and the Chamber of Secrets (Book 2),J. K. Rowling,1999,Scholastic,http://images.amazon.com/images/P/0439064864.0...,183
5092,254,0439064872,9,Harry Potter and the Chamber of Secrets (Book 2),J. K. Rowling,2000,Scholastic,http://images.amazon.com/images/P/0439064872.0...,183
5093,254,0439136350,9,Harry Potter and the Prisoner of Azkaban (Book 3),J. K. Rowling,1999,Scholastic,http://images.amazon.com/images/P/0439136350.0...,138
5094,254,0439136369,9,Harry Potter and the Prisoner of Azkaban (Book 3),J. K. Rowling,2001,Scholastic,http://images.amazon.com/images/P/0439136369.0...,138
5070,254,0380789019,0,Neverwhere,Neil Gaiman,1998,Avon,http://images.amazon.com/images/P/0380789019.0...,72
5074,254,0380973634,10,Neverwhere,Neil Gaiman,1997,William Morrow,http://images.amazon.com/images/P/0380973634.0...,72
5066,254,037582345X,9,"The Golden Compass (His Dark Materials, Book 1)",PHILIP PULLMAN,2002,Knopf Books for Young Readers,http://images.amazon.com/images/P/037582345X.0...,131
5187,254,0679879242,8,"The Golden Compass (His Dark Materials, Book 1)",PHILIP PULLMAN,1996,Knopf Books for Young Readers,http://images.amazon.com/images/P/0679879242.0...,131
5067,254,0375823468,9,"The Subtle Knife (His Dark Materials, Book 2)",PHILIP PULLMAN,2002,Knopf Books for Young Readers,http://images.amazon.com/images/P/0375823468.0...,83
5188,254,0679879250,0,"The Subtle Knife (His Dark Materials, Book 2)",Philip Pullman,1997,Alfred A. Knopf,http://images.amazon.com/images/P/0679879250.0...,83


In [213]:
max_ratings = final_ratings_df.groupby(['user_id', 'title'])['rating'].max().reset_index()
max_ratings

,user_id,title,rating
0,254,1984,9
1,254,A Bend in the Road,0
2,254,ANGELA'S ASHES,0
3,254,American Gods,10
4,254,Angela's Ashes (MMP) : A Memoir,0
...,...,...,...
57947,278418,Watership Down,0
57948,278418,Welcome to Temptation,0
57949,278418,Where the Heart Is (Oprah's Book Club (Paperba...,0
57950,278418,Where the Red Fern Grows,0


In [215]:
final_ratings_df.columns

Index(['user_id', 'title', 'rating', 'isbn', 'author', 'year_of_publication',
       'publisher', 'image_url_L', 'no_of_ratings'],
      dtype='object')

In [214]:
final_ratings_df = final_ratings_df.groupby(['user_id', 'title'], as_index=False).agg({
    'rating': 'max',
    'isbn': 'first',
    'author': 'first',
    'year_of_publication': 'first',
    'publisher': 'first',
    'image_url_L': 'first',
    'no_of_ratings': 'first'
})

In [216]:
final_ratings_df.shape

(57952, 9)

# Create a pivot table

In [217]:
book_pivot = final_ratings_df.pivot_table(columns='user_id', index='title', values='rating')

In [218]:
book_pivot

user_id,254,2276,2766,2977,3363,3757,4017,4385,6242,6251,...,274004,274061,274301,274308,274808,275970,277427,277478,277639,278418
title,,,,,,,,,,,,,,,,,,,,,
1984,9.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN
1st to Die: A Novel,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2nd Chance,NaN,10.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,0.0,NaN
4 Blondes,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
84 Charing Cross Road,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,10.0,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Year of Wonders,NaN,NaN,NaN,7.0,NaN,NaN,NaN,NaN,7.0,NaN,...,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN
You Belong To Me,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Zen and the Art of Motorcycle Maintenance: An Inquiry into Values,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,0.0,...,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN


The key distinction:

NaN → no data point exists (absence)

0.0 → a data point exists, and its value happens to be zero (a real recorded "no rating given")

In [219]:
book_pivot.shape

(703, 888)

### Convert the NaNs to 0s before converting to a sparse matrix

In [220]:
book_pivot.fillna(0, inplace=True)

In [221]:
book_pivot

user_id,254,2276,2766,2977,3363,3757,4017,4385,6242,6251,...,274004,274061,274301,274308,274808,275970,277427,277478,277639,278418
title,,,,,,,,,,,,,,,,,,,,,
1984,9.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1st to Die: A Novel,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2nd Chance,0.0,10.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4 Blondes,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
84 Charing Cross Road,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,10.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Year of Wonders,0.0,0.0,0.0,7.0,0.0,0.0,0.0,0.0,7.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
You Belong To Me,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Zen and the Art of Motorcycle Maintenance: An Inquiry into Values,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [222]:
from scipy.sparse import csr_matrix

In [223]:
book_sparse = csr_matrix(book_pivot)

In [224]:
book_sparse

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 14787 stored elements and shape (703, 888)>

Only 14,787 user-book rating pairs actually exist out of a possible 703 × 888 = 624,264 combinations.

In [239]:
sparsity = 1 - (14787 / (703 * 888))
print(f"{sparsity:.2%}")

97.63%


That works out to about 97.6% sparse — meaning only ~2.4% of all possible user-book cells have an actual rating. 
The other 97.6% are zeros (either genuinely unrated, or explicit-0 implicit ratings

In [225]:
type(book_sparse)

scipy.sparse._csr.csr_matrix

In [242]:
from sklearn.neighbors import NearestNeighbors

In [265]:
book_pivot.index.get_loc("Harry Potter and the Chamber of Secrets (Book 2)")

224

## 1. Test with current Euclidean-based model (already fitted)

In [ ]:
# book_pivot.iloc[237,:]

In [266]:
model = NearestNeighbors(algorithm='brute', metric='euclidean')
model.fit(book_sparse)

distance, suggestion = model.kneighbors(book_pivot.iloc[224,:].values.reshape(1,-1), n_neighbors=6)

print("Euclidean results:")
for i in suggestion[0]:
    print(book_pivot.index[i])
print("Distances:", distance[0])

Euclidean results:
Harry Potter and the Chamber of Secrets (Book 2)
Harry Potter and the Prisoner of Azkaban (Book 3)
Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Sorcerer's Stone (Book 1)
Exclusive
Jacob Have I Loved
Distances: [ 0.         68.92024376 69.76388751 72.96574539 77.64663547 78.14729682]


## 2. Now refit with cosine similarity

In [268]:
model_cosine = NearestNeighbors(algorithm='brute', metric='cosine')
model_cosine.fit(book_sparse)

distance_cos, suggestion_cos = model_cosine.kneighbors(book_pivot.iloc[224,:].values.reshape(1,-1), n_neighbors=6)

print("\nCosine results:")
for i in suggestion_cos[0]:
    print(book_pivot.index[i])
print("Distances:", distance_cos[0])


Cosine results:
Harry Potter and the Chamber of Secrets (Book 2)
Harry Potter and the Prisoner of Azkaban (Book 3)
Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Sorcerer's Stone (Book 1)
Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))
Harry Potter and the Order of the Phoenix (Book 5)
Distances: [1.11022302e-16 3.98252386e-01 4.34412211e-01 5.34821478e-01
 5.35667246e-01 6.43251541e-01]


In [269]:
book_pivot.index[224]

'Harry Potter and the Chamber of Secrets (Book 2)'

### Map the neighbor indices to titles (for both models):

In [270]:
print("Queried book:", book_pivot.index[224])

print("\nEuclidean neighbors:")
for i in suggestion[0]:
    print(book_pivot.index[i])

print("\nCosine neighbors:")
for i in suggestion_cos[0]:
    print(book_pivot.index[i])

Queried book: Harry Potter and the Chamber of Secrets (Book 2)

Euclidean neighbors:
Harry Potter and the Chamber of Secrets (Book 2)
Harry Potter and the Prisoner of Azkaban (Book 3)
Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Sorcerer's Stone (Book 1)
Exclusive
Jacob Have I Loved

Cosine neighbors:
Harry Potter and the Chamber of Secrets (Book 2)
Harry Potter and the Prisoner of Azkaban (Book 3)
Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Sorcerer's Stone (Book 1)
Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))
Harry Potter and the Order of the Phoenix (Book 5)


###  To see book + distance together (more useful for comparison):

In [271]:
print("Euclidean:")
for idx, dist in zip(suggestion[0], distance[0]):
    print(f"{book_pivot.index[idx]}  —  distance: {dist:.4f}")

print("\nCosine:")
for idx, dist in zip(suggestion_cos[0], distance_cos[0]):
    print(f"{book_pivot.index[idx]}  —  distance: {dist:.4f}")

Euclidean:
Harry Potter and the Chamber of Secrets (Book 2)  —  distance: 0.0000
Harry Potter and the Prisoner of Azkaban (Book 3)  —  distance: 68.9202
Harry Potter and the Goblet of Fire (Book 4)  —  distance: 69.7639
Harry Potter and the Sorcerer's Stone (Book 1)  —  distance: 72.9657
Exclusive  —  distance: 77.6466
Jacob Have I Loved  —  distance: 78.1473

Cosine:
Harry Potter and the Chamber of Secrets (Book 2)  —  distance: 0.0000
Harry Potter and the Prisoner of Azkaban (Book 3)  —  distance: 0.3983
Harry Potter and the Goblet of Fire (Book 4)  —  distance: 0.4344
Harry Potter and the Sorcerer's Stone (Book 1)  —  distance: 0.5348
Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))  —  distance: 0.5357
Harry Potter and the Order of the Phoenix (Book 5)  —  distance: 0.6433


In [272]:
final_model = NearestNeighbors(algorithm='brute', metric='cosine')
final_model.fit(book_sparse)

distance, suggestion = final_model.kneighbors(book_pivot.iloc[224,:].values.reshape(1,-1), n_neighbors=6)

In [273]:
distance

array([[1.11022302e-16, 3.98252386e-01, 4.34412211e-01, 5.34821478e-01,
        5.35667246e-01, 6.43251541e-01]])

In [274]:
suggestion

array([[224, 227, 225, 228, 229, 226]])

In [275]:
final_ratings_df[
    final_ratings_df['title'].str.contains('Harry Potter', case=False, na=False)
]

,user_id,title,rating,isbn,author,year_of_publication,publisher,image_url_L,no_of_ratings
15,254,Harry Potter and the Chamber of Secrets (Book 2),9,0439064864,J. K. Rowling,1999,Scholastic,http://images.amazon.com/images/P/0439064864.0...,183
16,254,Harry Potter and the Goblet of Fire (Book 4),9,0439139597,J. K. Rowling,2000,Scholastic,http://images.amazon.com/images/P/0439139597.0...,125
17,254,Harry Potter and the Order of the Phoenix (Boo...,0,043935806X,J. K. Rowling,2003,Scholastic,http://images.amazon.com/images/P/043935806X.0...,105
18,254,Harry Potter and the Prisoner of Azkaban (Book 3),9,0439136350,J. K. Rowling,1999,Scholastic,http://images.amazon.com/images/P/0439136350.0...,138
19,254,Harry Potter and the Sorcerer's Stone (Book 1),9,0590353403,J. K. Rowling,1998,Scholastic,http://images.amazon.com/images/P/0590353403.0...,78
...,...,...,...,...,...,...,...,...,...
57605,275970,Harry Potter and the Chamber of Secrets (Book 2),9,0439420105,J. K. Rowling,2002,Scholastic,http://images.amazon.com/images/P/0439420105.0...,183
57606,275970,Harry Potter and the Sorcerer's Stone (Book 1),9,043936213X,J. K. Rowling,2001,Scholastic,http://images.amazon.com/images/P/043936213X.0...,78
57692,277427,Harry Potter and the Chamber of Secrets (Book 2),0,0439064872,J. K. Rowling,2000,Scholastic,http://images.amazon.com/images/P/0439064872.0...,183
57693,277427,Harry Potter and the Prisoner of Azkaban (Book 3),0,0439136369,J. K. Rowling,2001,Scholastic,http://images.amazon.com/images/P/0439136369.0...,138


In [283]:
book_names = list(book_pivot.index)

# Saving the files

In [284]:
import pickle
pickle.dump(final_model, open('artifacts/final_model.pkl', 'wb'))
pickle.dump(final_ratings_df,open('artifacts/final_rating.pkl','wb'))
pickle.dump(book_pivot,open('artifacts/book_pivot.pkl','wb'))
pickle.dump(book_names,open('artifacts/book_names.pkl','wb'))

In [285]:
book_names

['1984',
 '1st to Die: A Novel',
 '2nd Chance',
 '4 Blondes',
 '84 Charing Cross Road',
 'A Bend in the Road',
 'A Case of Need',
 'A Child Called \\It\\": One Child\'s Courage to Survive"',
 'A Civil Action',
 'A Day Late and a Dollar Short',
 'A Fine Balance',
 'A Great Deliverance',
 'A Heartbreaking Work of Staggering Genius',
 'A Is for Alibi (Kinsey Millhone Mysteries (Paperback))',
 'A Lesson Before Dying (Vintage Contemporaries (Paperback))',
 'A Man in Full',
 'A Map of the World',
 'A Painted House',
 'A Patchwork Planet',
 'A Prayer for Owen Meany',
 'A Thin Dark Line (Mysteries &amp; Horror)',
 "A Thousand Acres (Ballantine Reader's Circle)",
 'A Time to Kill',
 "A Virtuous Woman (Oprah's Book Club (Paperback))",
 'A Walk to Remember',
 'A Widow for One Year',
 'A Wrinkle In Time',
 'A Wrinkle in Time',
 'A Year in Provence',
 "ANGELA'S ASHES",
 'About a Boy',
 'Absolute Power',
 'Acceptable Risk',
 'Accident',
 'Airframe',
 'All Around the Town',
 'All I Really Need to Kno

# Testing model

In [286]:
def recommend_book(book_name):
    book_id = np.where(book_pivot.index == book_name)[0][0]
    distance, suggestion = model.kneighbors(book_pivot.iloc[book_id,:].values.reshape(1,-1), n_neighbors=6 )
    
    for i in range(len(suggestion)):
            books = book_pivot.index[suggestion[i]]
            for j in books:
                if j == book_name:
                    print(f"You searched '{book_name}'\n")
                    print("The suggestion books are: \n")
                else:
                    print(j)

In [287]:
book_name = "2nd Chance"
recommend_book(book_name)

You searched '2nd Chance'

The suggestion books are: 

The Next Accident
Last Man Standing
Exclusive
Unspeakable
Standoff
